app/src/main/java/com/babylon/wallet/android/presentation/account/recover/AccountRecoveryViewModel.kt

```diff
@@ -0,0 +1,269 @@
+                        it.copy(selected = !it.selected)
+                    }
+                ).toPersistentList(),
+            )
+        }
+    }
+
+    fun startRecoveryScan() {
+        runningScan?.cancel()
+        runningScan = viewModelScope.launch {
+            state.value.recoveryFactorSource?.let { recoveryFS ->
+                recoverAccountsForFactorSourceUseCase(recoveryFS)
+                    .onSuccess { accounts ->
+                        if (isActive) {
+                            deriveStateFromRecoveredAccounts(accounts)
+                        }
+                    }
+                    .onFailure { failure ->
+                        _state.update { it.copy(uiMessage = UiMessage.ErrorMessage(failure)) }
+                        sendEvent(Event.CloseScan)
+                    }
+            }
+        }
+    }
+
+    fun onBackClick() {
+        viewModelScope.launch {
+            sendEvent(Event.OnBackClick)
+            _state.update { State() }
+        }
+    }
+
+    private fun deriveStateFromRecoveredAccounts(accounts: List<AccountWithOnLedgerStatus>) {
+        val allRecoveredAccounts = state.value.recoveredAccounts + accounts
+        val activeAccounts = allRecoveredAccounts.filter {
+            it.status == AccountWithOnLedgerStatus.Status.Active
+        }.map { it.account }.toPersistentList()
+        val maxActiveIndex = allRecoveredAccounts.indexOfLast { it.status == AccountWithOnLedgerStatus.Status.Active }
+        val inactiveAccounts = if (maxActiveIndex == -1) {
+            persistentListOf()
+        } else {
+            allRecoveredAccounts.subList(0, maxActiveIndex).filter { it.status == AccountWithOnLedgerStatus.Status.Inactive }
+                .map { it.account }
+                .toPersistentList()
+        }
+        _state.update { state ->
+            state.copy(
+                contentState = State.ContentState.ScanComplete,
+                recoveredAccounts = allRecoveredAccounts.toPersistentList(),
+                activeAccounts = activeAccounts,
+                inactiveAccounts = inactiveAccounts.map { Selectable(it) }.toPersistentList()
+            )
+        }
+    }
+
+    data class State(
+        val recoveryFactorSource: RecoveryFactorSource? = null,
+        val contentState: ContentState = ContentState.ScanInProgress,
+        val uiMessage: UiMessage? = null,
+        val recoveredAccounts: ImmutableList<AccountWithOnLedgerStatus> = persistentListOf(),
+        val interactionState: InteractionState? = null,
+        val activeAccounts: PersistentList<Network.Account> = persistentListOf(),
+        val inactiveAccounts: PersistentList<Selectable<Network.Account>> = persistentListOf(),
+        val isRestoring: Boolean = false
+    ) : UiState {
+
+        enum class ContentState {
+            ScanInProgress,
+            ScanComplete
+        }
+    }
+}
+
+sealed interface RecoveryFactorSource {
+    data class VirtualDeviceFactorSource(
```

Comment:
What's the `VirtualDeviceFactorSource` for ? How is that different than `RecoveryFactorSource.Device` ?

Reply:
it is virtual becuase I don't store it until whole flow is complete. This is only for onboarding